### ISSUE TITLE
" user is typing ... " doesn't appear after opening and closing the RHP

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#25486]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login with 2 users ( User A and User B ) in front of you at the same time.
2. Start a chat between the two users.
3. As User A click on settings to Open the RHP and then close it.
4. Now go to User B and start typing.
## Expected Result:
 " user is typing... " message should appear for User A while User B is typing.
## Actual Result: 
After opening and closing the RHP, the " user is typing ... " message doesn't appear for User A while User B is typing.
Note: Refreshing the page or leaving and returning to the chat resolves the issue.
Notes:
1. It doesn't matter which mode you switch to, whether from Most Recent to Focus or vice versa. The issue occurs after changing modes.
2. Refreshing the page or leaving and returning to the chat resolves the issue

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.55-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/954ac837-978d-477a-af3c-21d0a89fc8d8

[Screencast from 10-08-23 03_22_46 (1).webm](https://github.com/Expensify/App/assets/43996225/46c5a8db-1221-422c-bf85-8db337843624)

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Ahmed-Abdella
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1691544088203549

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

" user is typing ... " doesn't appear after changing the priority mode.

### What is the root cause of that problem?

The root cause of this issue is that we unsubscribe the typing status event in the cleanup function of this [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L140-L192) here

https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L184

And we don't re-subscribe the typing status event in this [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L140-L192) and in the other useEffect

https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L268-L278

also doesn't re-subscribe the event because the dependency array

https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L278

doesn't trigger the effect to re-run.

So, the typing status event is unsubscribed and won't get re-subscribed as long as this [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L140-L192) is triggered to re-run.

Note we explicitly call the cleanup function [here](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L141-L151) apart from being calling every re-render as the [official doc](https://react.dev/reference/react/useEffect#parameters) mentions.

### What changes do you think we should make in order to solve the problem?

To fix this issue, we should re-subscribe the typing status event if it's unsubscribed by the cleanup function.

To achieve it, 

1. We can mutate 

```js
didSubscribeToReportTypingEvents.current = false;
```

after the [line](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L184) to unsubscribe the event.

2. Add  `isReportFullyVisible` to the dependency array of the useEffect to subscribe the typing status event. We can change this line 

   https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L278

   to

   ```js
   }, [props.report, didSubscribeToReportTypingEvents, reportID, isReportFullyVisible]);
   ```

   We can also remove `didSubscribeToReportTypingEvents` from the deps array as changing value of `useRef` doesn't trigger a re-render.

### What alternative solutions did you explore? (Optional)

We can also merge the two useEffects,  [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L140-L192) and [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L268-L278), into one useEffect.

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

" user is typing ... " doesn't appear after changing the priority mode.

### What is the root cause of that problem?

The root cause of this issue is that we unsubscribe the typing status event in the cleanup function of this [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L140-L192) here

https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L184

And we don't re-subscribe the typing status event in this [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L140-L192) and in the other useEffect

https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L268-L278

also doesn't re-subscribe the event because the dependency array

https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L278

doesn't trigger the effect to re-run.

So, the typing status event is unsubscribed and won't get re-subscribed as long as this [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L140-L192) is triggered to re-run.

Note we explicitly call the cleanup function [here](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L141-L151) apart from being calling every re-render as the [official doc](https://react.dev/reference/react/useEffect#parameters) mentions.

### What changes do you think we should make in order to solve the problem?

To fix this issue, we should re-subscribe the typing status event if it's unsubscribed by the cleanup function.

To achieve it, 

1. We can mutate 

```js
didSubscribeToReportTypingEvents.current = false;
```

after the [line](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L184) to unsubscribe the event.

2. Add  `isReportFullyVisible` to the dependency array of the useEffect to subscribe the typing status event. We can change this line 

   https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L278

   to

   ```js
   }, [props.report, didSubscribeToReportTypingEvents, reportID, isReportFullyVisible]);
   ```

   We can also remove `didSubscribeToReportTypingEvents` from the deps array as changing value of `useRef` doesn't trigger a re-render.

### What alternative solutions did you explore? (Optional)

We can also merge the two useEffects,  [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L140-L192) and [useEffect](https://github.com/Expensify/App/blob/0f83752c0e77ec9c6302e7a00915cfc7a772ff3b/src/pages/home/report/ReportActionsView.js#L268-L278), into one useEffect.

**Comment by kavimuru** [comment]
@Pujan92 will this be fixed when this deployed to staging? 
https://github.com/Expensify/App/pull/18637 ?

**Comment by Pujan92** [comment]
Yes @kavimuru , some follow up change we can do which I have mentioned [here](https://github.com/Expensify/App/pull/18637#issuecomment-1683413113).
cc: @gedu 

<details>
<summary>Cause for this issue which won't be reproducible now</summary>

## Proposal

### Please re-state the problem that we are trying to solve in this issue.
'User is typing' isn't shown once we lose focus from the report(not specifically required to change priority mode)

### What is the root cause of that problem?
Currently, we are unsubscribing the Typing Pusher event at the wrong place. Cleanup function will be called whenever the dep gets changed but Typing event needs to be unsubscribed when the component unmounts. So the current issue is whenever the report focus gets lost it will trigger this cleanup and unsubscribe the typing event. Seems it persists after this [PR](https://github.com/Expensify/App/pull/23729) 
cc @hannojg 

https://github.com/Expensify/App/blob/77ac9d18f3bc122c10b1baa7f566f37e51b3de9a/src/pages/home/report/ReportActionsView.js#L180-L192

### What changes do you think we should make in order to solve the problem?
We need to move `Report.unsubscribeFromReportChannel(reportID);` here by creating the cleanup function which will unsubscribe typing event on unmount.

https://github.com/Expensify/App/blob/77ac9d18f3bc122c10b1baa7f566f37e51b3de9a/src/pages/home/report/ReportActionsView.js#L135-L138
</details>

**Comment by adelekennedy** [comment]
It seems like this was fixed by #18637 - let's close this for now
